finality-aleph/src/testing/data_store.rs

```diff
@@ -0,0 +1,645 @@
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    test_handler.finalize_block(&blocks[MAX_DATA_BRANCH_LEN + 8].hash());
+
+    let _message = test_handler
+        .assert_message_out("Did not receive message from Data Store")
+        .await;
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn sends_block_request_on_missing_block() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10)
+        .await;
+    let blocks_branch = blocks[0..1].to_vec();
+    let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+    test_handler.send_data(test_data.clone());
+
+    test_handler
+        .assert_no_message_out("Data Store let through a message with not finalized ancestor")
+        .await;
+
+    let requested_block = timeout(TIMEOUT_SUCC, test_handler.next_block_request())
+        .await
+        .expect("Did not receive block request from Data Store");
+    assert_eq!(requested_block.hash, blocks[0].hash());
+
+    test_handler.import_branch(blocks, false).await;
+
+    let _message = test_handler
+        .assert_message_out("Did not receive message from Data Store")
+        .await;
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn does_not_send_requests_when_no_block_missing() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    for i in 1..=MAX_DATA_BRANCH_LEN {
+        let blocks_branch = blocks[0..(i as usize)].to_vec();
+        let test_data: TestData = vec![aleph_data_from_blocks(blocks_branch)];
+        test_handler.send_data(test_data.clone());
+    }
+    assert!(
+        timeout(TIMEOUT_FAIL, test_handler.next_block_request())
+            .await
+            .is_err(),
+        "Data Store is sending block requests without reason"
+    );
+
+    test_handler.exit();
+    data_store_handle.await.unwrap();
+}
+
+#[tokio::test]
+async fn message_with_genesis_block_does_not_get_through() {
+    let (task_handle, mut test_handler) = prepare_data_store();
+    let data_store_handle = tokio::spawn(task_handle);
+    let genesis_hash = test_handler.genesis_hash();
+
+    let _blocks = test_handler
+        .build_and_import_branch_upon(&genesis_hash, MAX_DATA_BRANCH_LEN * 10, false)
+        .await;
+
+    let test_data: TestData = vec![aleph_data_from_headers(vec![test_handler.get_header_at(0)])];
```

Comment:
It seems that a message containing something more than genesis block would be a more general testcase